de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/DraftInventoryLinesCreator.java

```diff
@@ -0,0 +1,152 @@
+{
+	IDocumentBL documentBL = Services.get(IDocumentBL.class);
+	IInventoryDAO inventoryDAO = Services.get(IInventoryDAO.class);
+	IHandlingUnitsBL handlingUnitsBL = Services.get(IHandlingUnitsBL.class);
+
+	I_M_Inventory inventoryRecord;
+
+	Map<HuId, I_M_InventoryLine> inventoryLinesByHU;
+	final Set<Integer> seenLocatorIs = new HashSet<>();
+
+	@NonFinal
+	long countInventoryLines = 0;
+
+	@Builder
+	private DraftInventoryLinesCreator(
+			@NonNull final I_M_Inventory inventoryRecord,
+			@NonNull final HUsForInventoryStrategy strategy)
+	{
+		Check.errorUnless(
+				documentBL.issDocumentDraftedOrInProgress(inventoryRecord),
+				"the given inventory record needs to be in status 'DR' or 'IP', but is in status={}; inventoryRecord={}",
+				inventoryRecord.getDocStatus(), inventoryRecord);
+		this.inventoryRecord = inventoryRecord;
+
+		// get existing lines' HuIds
+		this.inventoryLinesByHU = inventoryDAO
+				.retrieveLinesForInventoryId(inventoryRecord.getM_Inventory_ID(), I_M_InventoryLine.class)
+				.stream()
+				.filter(line -> line.getM_HU_ID() > 0)
+				.collect(GuavaCollectors.toImmutableMapByKey(line -> HuId.ofRepoId(line.getM_HU_ID())));
+
+		// create/update new lines
+		final Iterator<I_M_HU> hus = strategy.streamHus().iterator();
+		while (hus.hasNext())
+		{
+			final I_M_HU hu = hus.next();
+			seenLocatorIs.add(hu.getM_Locator_ID());
+			
+			if (!strategy.match(seenLocatorIs.size()))
+			{
+				return;
+			}
+
+			countInventoryLines = countInventoryLines + createUpdateInventoryLines(hu).count();
+		}
+	}
+
+	private Stream<I_M_InventoryLine> createUpdateInventoryLines(@NonNull final I_M_HU hu)
+	{
+		return handlingUnitsBL
+				.getStorageFactory()
+				.streamHUProductStorages(hu)
+				.filter(huProductStorage -> !huProductStorage.isEmpty())
+				.map(this::createOrUpdateInventoryLine);
+	}
+
+	private I_M_InventoryLine createOrUpdateInventoryLine(@NonNull final IHUProductStorage huProductStorage)
+	{
+		final I_M_InventoryLine inventoryLine;
+		final I_M_HU hu = huProductStorage.getM_HU();
+
+		final HuId huId = HuId.ofRepoId(hu.getM_HU_ID());
+
+		if (inventoryLinesByHU.containsKey(huId))
+		{
+			// update line
+			inventoryLine = inventoryLinesByHU.get(huId);
+		}
+		else
+		{
+			// create line
+			inventoryLine = InterfaceWrapperHelper.newInstance(I_M_InventoryLine.class);
+			inventoryLine.setM_Inventory(inventoryRecord);
+			inventoryLine.setAD_Org_ID(inventoryRecord.getAD_Org_ID());
+		}
+
+		inventoryLine.setM_AttributeSetInstance_ID(0);
+		inventoryLine.setM_HU_ID(hu.getM_HU_ID());
+		inventoryLine.setM_HU_PI_Item_Product(null); // TODO
+		inventoryLine.setQtyTU(BigDecimal.ZERO); // TODO
+
+		inventoryLine.setM_Locator_ID(hu.getM_Locator_ID());
+		inventoryLine.setM_Product_ID(huProductStorage.getM_Product_ID());
+		inventoryLine.setC_UOM(huProductStorage.getC_UOM());
+
+		inventoryLine.setQtyBook(huProductStorage.getQty());
+		inventoryLine.setQtyCount(huProductStorage.getQty());
+
+		saveRecord(inventoryLine);
```

Comment:
i would prefer to introduce IInventoryDAO.save method and call it here.
I would avoid calling InterfaceWrapperHelper.save, delete etc methods outside of a Repository/DAO.